xtask/src/cargo.rs

```diff
@@ -0,0 +1,211 @@
+
+    /// Update the version of this crate in dependant crates' manifest, with the given version
+    /// prefix.
+    pub fn update_dependants(&self, packages: &[Package]) -> Result<()> {
+        for package in
+            packages.iter().filter(|p| p.dependencies.iter().any(|d| d.name == self.name))
+        {
+            println!("Updating dependency in {} crate…", package.name);
+
+            let mut document = read_file(&package.manifest_path)?.parse::<Document>()?;
+
+            for dependency in package.dependencies.iter().filter(|d| d.name == self.name) {
+                let version = if dependency.req.is_exact() {
+                    format!("={}", self.version,)
+                } else {
+                    self.version.to_string()
+                };
+
+                match dependency.kind {
+                    Some(DependencyKind::Dev) => {
+                        document["dev-dependencies"][&self.name]["version"] =
+                            value(version.as_str());
+                    }
+                    None => {
+                        document["dependencies"][&self.name]["version"] = value(version.as_str());
+                    }
+                    _ => {}
+                }
+            }
+
+            if package
+                .dependencies
+                .iter()
+                .any(|p| p.name == self.name && p.kind == Some(DependencyKind::Dev))
+            {}
+
+            write_file(&package.manifest_path, document.to_string())?;
+        }
+
+        Ok(())
+    }
+
+    /// Update the changelog for the release of the given version. Returns the changes for the
+    /// version.
+    pub fn update_changelog(&self) -> Result<String> {
+        let mut changelog_path = self.manifest_path.clone();
+        changelog_path.set_file_name("CHANGELOG.md");
+
+        let changelog = read_file(&changelog_path)?;
+
+        let title_start = match changelog
+            .find(&format!("# {}\n", self.version))
+            .or_else(|| changelog.find(&format!("# {} (unreleased)\n", self.version)))
+            .or_else(|| changelog.find("# [unreleased]\n"))
+        {
+            Some(p) => p,
+            None => {
+                return Err("Could not find version title in changelog".into());
+            }
+        };
+
+        let changes_start = match changelog[title_start..].find('\n') {
+            Some(p) => title_start + p + 1,
+            None => {
+                return Err("Could not find end of version title in changelog".into());
+            }
+        };
+
+        let changes_end = match changelog[changes_start..].find("\n# ") {
+            Some(p) => changes_start + p,
+            None => changelog.len(),
+        };
+
+        let changes = match changelog[changes_start..changes_end].trim() {
+            s if s.is_empty() => "No changes for this version",
+            s => s,
+        };
+
+        let changelog = format!(
+            "{}# [unreleased]\n\n# {}\n\n{}\n{}",
```

Comment:
I'm not entirely sure on this. So this potentially adds `# [unreleased]` somewhere else than at the start of the file? What for?

Reply:
This was in case there is some "header" text in the changelog, to be more flexible. But I guess that's useless in our case.